compiler/front_end/module_ir.py

```diff
@@ -153,11 +153,11 @@ def _really_build_ir(parse_tree, used_productions):
         ]
         used_productions.add(parse_tree.production)
         result = _handlers[parse_tree.production](*parsed_children)
-        if parse_tree.source_location is not None:
-            if result.source_location:
-                ir_data_utils.update(result.source_location, parse_tree.source_location)
+        if parse_tree.source_location:
+            if isinstance(result, tuple):
+                result = result._replace(source_location=parse_tree.source_location)
             else:
-                result.source_location = ir_data_utils.copy(parse_tree.source_location)
+                result.source_location = parse_tree.source_location
```

Comment:
Yeah, more or less the point of this change is to make it so that it is never necessary to deep copy `SourceLocation` -- it's frozen, so it doesn't matter if multiple nodes point to the same `SourceLocation` instance.  About half of the time in `_copy()` was just copying `Location`.

It would be (relatively) cheap to special case `update` and `copy` to check if their arguments are `tuple` (specifically checking for anything created by `namedtuple()` is harder), but I don't really see the point?  Right now, they just raise an exception if you give them something that isn't an IR dataclass, so any mistakes will be caught quickly.

Reply:
Yeah I think it's fine as-is, it just took me a second to grok why we weren't using the standard methods for this anymore.